src/HierarchicalCachePoolTrait.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace Cache\Hierarchy;
+
+use Cache\Adapter\Common\Exception\InvalidArgumentException;
+use Cache\Taggable\TaggablePoolInterface;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+trait HierarchicalCachePoolTrait
+{
+    /**
+     * A temporary cache for keys
+     *
+     * @type array
+     */
+    private $keyCache = [];
+
+    /**
+     * Get a value form the store. This must not be an PoolItemInterface.
+     *
+     * @param string $key
+     *
+     * @return string|null
+     */
+    abstract protected function getValueForStore($key);
+
+
+    /**
+     * Get a key to use with the hierarchy. If the key does not start with HierarchicalPoolInterface::SEPARATOR
+     * this will return an unalterered key. This function supports a tagged key. Ie "foo:bar".
+     *
+     * @param string $key The original key
+     * @param string &$pathKey A cache key for the path. If this key is changed everything beyond that path is changed.
+     *
+     * @return string
+     */
+    protected function getHierarchyKey($key, &$pathKey = null)
+    {
+        $hierarchyKey = new HierarchyKeyWithTags(
+          new HierarchyKey($key),
+          TaggablePoolInterface::TAG_SEPARATOR
+        );
+
+        if (!$hierarchyKey->isValid()) {
+            return $key;
```

Comment:
I think you're a bit paranoid about classes and performance. :P
But I could probably do some performance testing.

Talking performance I feel more uneasy about having this as a trait with no option to opt out of this functionality.
Using a wrapper class would solve that in a better way and make sure you're actually using it.

But then again for users who just want to instantiate one class and doesn't care another class could be used to do the wrapping on the inside.